src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -0,0 +1,483 @@
+
+    # make a log record if the DID + Rucio account tuple does not have any existent rules
+    if not rules:
+        msg = f"inactive task {uuid}, container: {pname} for Rucio account {rucioAccount}"
+        msg += " does not have any existing rules, proceed without update"
+        logger.info(msg)
+
+    # persist an up-to-date version of the pileup data structure in MongoDB
+    if modify:
+        logger.info(f"inactive task {uuid}, update {pname}")
+        mgr.updatePileup(doc)
+        msg = f"update pileup {pname}"
+        report.addEntry('inactive', uuid, msg)
+
+
+def activeTask(doc, spec):
+    """
+    Perform single active task over provided MSPileup document
+
+    :param doc: MSPileup document
+    :param spec: task spec dict
+    """
+    mgr = spec['manager']
+    uuid = spec['uuid']
+    logger = spec['logger']
+    rucioClient = spec['rucioClient']
+    rucioAccount = spec['rucioAccount']
+    report = spec['report']
+    marginSpace = spec['marginSpace']
+    nodeUsage = spec['nodeUsage']
+
+    # extract relevant part of our pileup document we'll use in our logic
+    expectedRSEs = set(doc['expectedRSEs'])
+    currentRSEs = set(doc['currentRSEs'])
+    pileupSize = doc['pileupSize']
+    pname = doc['pileupName']
+    inputArgs = {'scope': 'cms', 'name': pname, 'account': rucioAccount}
+    kwargs = {'scope': 'cms', 'account': rucioAccount}
+    modify = False
+    localCopyOfCurrentRSEs = list(currentRSEs)
+
+    # if expectedRSEs is different than currentRSEs, then further data placement is required
+    # it's possible that data removal is required!
+    if expectedRSEs != currentRSEs:
+        msg = f"active task {uuid}"
+        msg += f", further data placement required for pileup name: {pname},"
+        msg += f" with expectedRSEs: {expectedRSEs} and data currently available at: {currentRSEs}"
+        logger.info(msg)
+
+        # get list of replication rules for our scope, pileup name and account
+        if isinstance(rucioClient, Client):
+            rules = list(rucioClient.list_replication_rules(inputArgs))
+        else:
+            rules = rucioClient.listDataRules(pname, **kwargs)
+        # for each rse_expression in rules get list of rses
+        for rdoc in rules:
+            if isinstance(rucioClient, Client):
+                rses = [r['rse'] for r in rucioClient.list_rses(rdoc['rse_expression'])]
+            else:
+                rses = rucioClient.evaluateRSEExpression(rdoc['rse_expression'])
+            rdoc['rses'] = rses
+
+        # for each rule matching the DID + Rucio account
+        for rdoc in rules:
+            rid = rdoc['id']
+            rses = rdoc['rses']
+            for rse in rses:
+                # if rule RSE is not in expectedRSEs, then this rule needs to be deleted
+                if rse not in expectedRSEs:
+                    # upon successful rule deletion, also remove the RSE name from currentRSEs
+                    if isinstance(rucioClient, Client):
```

Comment:
It has not been mentioned in the requirements, but let me write it here before we forget it.

We should have a mechanism to run the "rule deletion" in dry-run mode, such that we don't really delete any rule, but just log that it would have been deleted.